modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultHashIndex.java

```diff
@@ -0,0 +1,469 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.h2.disk;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIO;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIOFactory;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.h2.value.Value;
+import org.h2.value.ValueRow;
+
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+import static java.nio.file.StandardOpenOption.READ;
+import static java.nio.file.StandardOpenOption.WRITE;
+
+/**
+ * Disk-based hash index for the sorted external result.
+ * Open addressing hashing scheme is used.
+ */
+@SuppressWarnings("TypeMayBeWeakened")
+public class ExternalResultHashIndex implements AutoCloseable {
```

Comment:
I was thinking about it when designed the solution. Two main reasons why I decided to  introduce additional index `ExternalResultHashIndex`:

1. It was undesirable to change H2 native code more than it is needed for solving the current tasks. It is caused by the difficulties of ensuring that H2 code works as expected. If we break something in H2 code, it is very difficult to detect the problem. So, I decided to implement `org.h2.result.ResultExternal` interface without changing it at all and with a minors changes to 'H2ManagedLocalResult' and other classes which are used by H2 engine. And in this case some methods like `org.h2.result.ResultExternal#addRow` can not be implemented without fast duplicates detection which is done by  `ExternalResultHashIndex`.
2. Methods like 'org.h2.result.ResultExternal#removeRow'  and `org.h2.result.ResultExternal#contains` should work fast, and `ExternalResultHashIndex` contributes to. If we use LSM style for finding duplicates, we should check all sorted chunks on the diskm which not be as fast as  `ExternalResultHashIndex`. So it can lead to performance issues for some types of queries as well.

Reply:
1. As I understand you need to guarantee that org.h2.result.ResultExternal#next wouldn't return duplication but you must not do it exactly in org.h2.result.ResultExternal#addRow
2. org.h2.result.ResultExternal#removeRow it has at least two fetches from disk so it certainly will be relatively slow.
LSM style does the same thing which you already did in your SortedExternalResult#next, doesn't it? You need just added one more check that 'next != current'.

My point of view - you should do as simpler as possible if you don't have any proofs to do more complicated. 
LSM way will be slower only in one scenario when we have a lot of same keys which fall into different chunks but how real is this scenario is? I mean now you have class ExternalResultHashIndex which complex enough and involve one more IO operation with the rollover of file. I concerned that it can lead to new bugs(usually this lead).
Anyway, it depends on your expected load profile and maybe it makes sense to do benchmark on both approach(as I understand the implementation of LSM way should not be difficult on current codebase).

Reply:
I only now figured out your comment about org.h2.result.ResultExternal#contains so it is a good point. Maybe hash table it is better variant. I try to figure out it.

Reply:
@akalash, are you *definitely* sure that LSM approach fits significantly better here? Are the some crucial flaws in a current design? Are there any blockers to come up with a better solution some time later? I suggest to proceed with a current design if it is good enough.

Reply:
According to the requirement of using 'contains' method current solution is good enough for me. Although I still have some concerns about it cause it makes code more difficult but currently, I don't have any other solution.